File: android/app/src/main/java/com/hcjeong/forestix/data/cruise/ProjectRepository.kt
// Port of iOS Persistence/Repositories/ProjectRepository.swift.
// Spec §8 + REQ-PRJ-001 (CRUD persists across app restarts).
//
// The iOS protocol's throwing sync functions become suspend functions; the
// Room implementation (`RoomProjectRepository`) plays the role of
// `CoreDataProjectRepository`. `observeList()` is the Compose-side extra —
// a cold Flow the project list screen collects like an @FetchRequest.

package com.hcjeong.forestix.data.cruise

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.UUID

interface ProjectRepository {
    suspend fun create(project: Project): Project
    suspend fun read(id: UUID): Project?
    suspend fun update(project: Project): Project
    suspend fun delete(id: UUID)
    suspend fun list(): List<Project>

    /// Live view of `list()` (newest-first) for Compose screens.
    fun observeList(): Flow<List<Project>>
}

class RoomProjectRepository(private val dao: ProjectDao) : ProjectRepository {

    override suspend fun create(project: Project): Project {
        dao.upsert(ProjectMapper.apply(project))
        return project
    }

    override suspend fun read(id: UUID): Project? =
        dao.byId(id)?.let(ProjectMapper::toStruct)

    override suspend fun update(project: Project): Project {
        dao.byId(project.id) ?: throw CruiseDataError.NotFound(project.id.toString())
        dao.upsert(ProjectMapper.apply(project))
        return project
    }

    override suspend fun delete(id: UUID) {
        if (dao.deleteById(id) == 0) throw CruiseDataError.NotFound(id.toString())
    }

    override suspend fun list(): List<Project> =
        dao.list().map(ProjectMapper::toStruct)

    override fun observeList(): Flow<List<Project>> =
        dao.observeList().map { rows -> rows.map(ProjectMapper::toStruct) }
}

File: TimberCruisingApp/Persistence/Repositories/ProjectRepository.swift
// Spec §8 + REQ-PRJ-001 (CRUD persists across app restarts).

import Foundation
import CoreData
import Models

public protocol ProjectRepository {
    func create(_ project: Project) throws -> Project
    func read(id: UUID) throws -> Project?
    func update(_ project: Project) throws -> Project
    func delete(id: UUID) throws
    func list() throws -> [Project]
}

public final class CoreDataProjectRepository: ProjectRepository {
    private let stack: CoreDataStack
    public init(stack: CoreDataStack) { self.stack = stack }

    public func create(_ project: Project) throws -> Project {
        try performWrite(stack: stack) { ctx in
            let e = try insert(ProjectEntity.self, entityName: "ProjectEntity", in: ctx)
            ProjectMapper.apply(project, to: e)
            return project
        }
    }

    public func read(id: UUID) throws -> Project? {
        try performRead(stack: stack) { ctx in
            guard let e = try fetchOne(ProjectEntity.self, entityName: "ProjectEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { return nil }
            return try ProjectMapper.toStruct(e)
        }
    }

    public func update(_ project: Project) throws -> Project {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(ProjectEntity.self, entityName: "ProjectEntity",
                                       keyPath: "id", value: project.id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: project.id.uuidString) }
            ProjectMapper.apply(project, to: e)
            return project
        }
    }

    public func delete(id: UUID) throws {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(ProjectEntity.self, entityName: "ProjectEntity",
                                       keyPath: "id", value: id as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: id.uuidString) }
            ctx.delete(e)
        }
    }

    public func list() throws -> [Project] {
        try performRead(stack: stack) { ctx in
            let sort = [NSSortDescriptor(key: "createdAt", ascending: false)]
            let entities = try fetchMany(ProjectEntity.self, entityName: "ProjectEntity",
                                         sort: sort, in: ctx)
            return try entities.map(ProjectMapper.toStruct)
        }
    }
}
